refactor: change internal msg format

luke-biel opened the pull request.

Kononnable commented on utils/src/message_types.rs:
```diff
@@ -1,21 +1,41 @@
+pub struct BorrowedInsertMessage<'a> {
     pub object_id: Uuid,
     pub schema_id: Uuid,
+    #[serde(default = "default_timestamp")]
```
Will lead to logic error which are hard to find. This should be set explicitly.

luke-biel replied:
Then DataRouter will require timestamps. This is now shared between DR and CS. I could make it optional, or '-1' by default. I don't know however if it's better. What do you think?

USSAMOH2 replied:
Between those two alternatives, I'd think that making it into a `pub timestamp: Option<i64>` makes it safer.

Kononnable replied:
>Then DataRouter will require timestamps. 

I say this shows need for another structure. Data router input format(client data) can't have this field - it's internal to CDL(in theory we could just overwrite its value but its very unclean). If we allow users to use it we would lock ourselves to current format of data(timestamp) which is likely to change in the future. After some time we would get questions on why something doesn't work. We don't support inserting historical data at this point and people would try to use it as such and probably broke their CDL data in some future version.

luke-biel replied:
Ok, this was my misconception about cdl's timestamp. This needs separate type for DataRouter, since DR isn't supposed to receive timestamps.

Kononnable commented on examples/deploy/docker-compose.yml:
```diff
@@ -134,7 +135,7 @@ services:
       POSTGRES_PORT: 5432
       POSTGRES_DBNAME: postgres
-      POSTGRES_SCHEMA: public
+      POSTGRES_SCHEMA: cdl
```
Was this intended or is it some change which is applicable only to your local environment(I have no idea if schema would be created automatically)?

luke-biel replied:
It's intended. Scripts provision this schema, example got out of sync.

Kononnable commented on command-service/src/output/victoria_metrics/mod.rs:
```diff
@@ -151,14 +153,13 @@ mod tests {
 
         #[test_case("{}"                => matches Err(Error::EmptyFields))]
         #[test_case("{ \"y01\": {} }"   => matches Err(Error::InvalidFieldType))]
-        #[test_case("y00=32q,y01=14.0f" => matches Err(Error::DataIsNotValidJson(_)))]
```
Removing test case in refactor PR isn't a good idea.

luke-biel replied:
It's not applicable anymore. RawValue cannot be an invalid json, str could.

Kononnable replied:
It's a refactor so errors/tests don't get obsolete. The only possibility is if this error was redundant in the first place(code could never be executed because of logic conditions, technically value could be just unwrapped).

Kononnable replied:
Or the other way around - in this PR we skip this error because of unwraping
`data: serde_json::from_str(&self.data.get()).expect("RawValue wasn't valid json Value"),`

luke-biel replied:
```        let event: BorrowedInsertMessage =
            serde_json::from_str(json).map_err(Error::PayloadDeserializationFailed)?;
``` is a place where this error will be caught now.
This refactor removed one unnecessary conversion. Earlier we checked if json was ok on deserializaton and then on serialization. Propagating type `InternalMessage` and RawValue means it is checked only on deserialization, and RawValue guarantees it holds a valid json. `build_line_protocol` had `payload: [u8]` argument, now accepts `payload: RawValue`. Slice could contain errors (in theory), RawValue can't. This function cannot, literally, fail here. Test is obsolete, because function changed.

Kononnable reviewed (changes requested):
What is the reasoning behind those changes? It's not clear to me why we should have both cases.

luke-biel commented:
> What is the reasoning behind those changes? It's not clear to me why we should have both cases.

I don't understand which `both cases` are you referring to. If it's Owned and Borrowed message - it was already there. CS had Owned message which was called GenericMessage. I replaced it with common construct, and removed one unnecessary allocation in the process.

USSAMOH2 commented on command-service/src/report/mod.rs:
```diff
@@ -1,37 +1,38 @@
+    pub fn with_message_body(self, msg: &BorrowedInsertMessage) -> Box<dyn Reporter> {
         match self {
             ReportSender::Full(config) => Box::new(FullReportSender {
                 producer: config.producer.clone(),
```
If this consumes `self`, then we probably don't need to `clone` these.

Kononnable commented on data-router/src/main.rs:
```diff
@@ -38,6 +39,14 @@ struct Config {
 }
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DataRouterInsertMessage<'a> {
```
I think this should go back to utils. Even if it's used only here it would be easier to find common message types when they're in one place.

luke-biel replied:
done

Kononnable reviewed (commented):
I don't like how FullReportSender is implemented(or at least its name compared to what it does), not sure if `BorrowedInsertMessage` and `OwnedInsertMessage` could be handled better(seems like that but don't see a concrete solution yet).
Anyway those issues should not block that PR.

Only the one more comment about moving `DataRouterInsertMessage` should be addressed, otherwise LGTM.

luke-biel commented:
> I don't like how FullReportSender is implemented(or at least its name compared to what it does), not sure if `BorrowedInsertMessage` and `OwnedInsertMessage` could be handled better(seems like that but don't see a concrete solution yet).
> Anyway those issues should not block that PR.
> 
> Only the one more comment about moving `DataRouterInsertMessage` should be addressed, otherwise LGTM.

I introduced NullObjectPattern because I could't come up with anything better that would not allocate memory when Reporting is disabled (coz payload has to be copied, for DE). Maybe we can hide that behind a feature? I'm just not sure whether that's the way.
